tests/transport/test_send_recv.cpp

```diff
@@ -184,63 +177,76 @@ auto test_ring_send_recv_cb_disown(CommType& comm, MsgType &&rmsg, MsgType &&sms
+    test_ring_send_recv_cb_disown< message_factory<gridtools::ghex::tl::message_buffer<>> >(comm, sizeof(int));
+    test_ring_send_recv_cb_disown< message_factory<gridtools::ghex::tl::shared_message_buffer<>> >(comm, sizeof(int));
+    test_ring_send_recv_cb_disown< message_factory<msg_type> >(comm, sizeof(int));
 }
 
+struct recursive_functor {
+    communicator_type& comm;
+    int rank;
+    int tag;
+    typename communicator_type::request_cb& rreq;
+    std::function<void(msg_type&, int, int)> func;
+    // we need a counter here so we can handle immediate callbacks
+    int counter = 0;
+
+    void operator()(msg_type m, int r, int t) {
+        func(m,r,t);
+        counter = 0;
+        auto rr = comm.recv(std::move(m), rank, tag, *this);
+        if (counter == 0) {
+            ++counter;
+            rreq = std::move(rr);
+        }
+    }
+};
 
-
-template<typename MsgType, typename CommType>
-auto test_ring_send_recv_cb_resubmit(CommType& comm, MsgType &&rmsg, MsgType &&smsg) 
+template<typename Factory, typename CommType>
+auto test_ring_send_recv_cb_resubmit(CommType& comm, std::size_t buffer_size) 
 {
     gridtools::ghex::timer timer;
     int *data_ptr;
     int rank = comm.rank();
     int size = comm.size();
     int speer_rank = (rank+1)%size;
-    int rpeer_rank = (rank-1)%size;
-    if(rpeer_rank<0) rpeer_rank = size-1;
+    int rpeer_rank = (rank+size-1)%size;
+    
+    auto smsg = Factory::make(buffer_size);
+    auto rmsg = Factory::make(buffer_size);
 
     timer.tic();    
-    communicator_type::request_cb_type rreq;
     volatile int received = 0;
-    volatile int sent = 0;
-    auto send_callback = [&](communicator_type::message_type, int, int) {sent++;};
 
-    std::function<void(communicator_type::message_type rmsg, int, int)> recv_callback;
-    recv_callback = [&](communicator_type::message_type rmsg, int, int) 
+    typename communicator_type::request_cb rreq;
+    auto recv_callback = [&received,rpeer_rank](communicator_type::message_type& rmsg, int, int) 
         {
             received++;
             int *data_ptr = reinterpret_cast<int*>(rmsg.data());
             EXPECT_TRUE(*data_ptr == rpeer_rank);
-            rreq = comm.recv(std::move(rmsg), rpeer_rank, 1, recv_callback);
+            *data_ptr = -1;
         };
+    auto recursive_recv_callback = recursive_functor{comm,rpeer_rank,1,rreq,recv_callback};
 
     data_ptr = reinterpret_cast<int*>(smsg.data());
     *data_ptr = rank;
 
-    rreq = comm.recv(rmsg, rpeer_rank, 1, recv_callback);
+
+    rreq = comm.recv(rmsg, rpeer_rank, 1, recursive_recv_callback);
     for(int i=0; i<NITERS; i++){
-        comm.send(smsg, speer_rank, 1, send_callback);
-        while(received<=i || sent<=i) comm.progress();
+        comm.send(smsg, speer_rank, 1).wait();
+        while(received<=i) comm.progress();
     }
 
-    EXPECT_TRUE(received==NITERS && sent==NITERS);
+    EXPECT_TRUE(received==NITERS);
     EXPECT_TRUE(rreq.cancel());
 
```

Comment:
we need a `comm.barrier()` after this line to make sure all ranks send/recv the correct number of requests, and cancel the last request before the peer manages to submit a matching send. Otherwise we screw up the order and hang :)